### PR TITLE
[EPIC] Parse args, correct frames order

### PR DESCRIPTION
**Pull request by talyguryn**
**was** and **now**

image

### CODE REVIEW

File: src/Util/Stacktrace.php [hunk 1]
```diff
@@ -35,18 +35,32 @@ public static function buildStack(Throwable $exception): array
...
+        $newStack = [];
+
         /**
          * Get real exception position
          */
         $errorPosition = [
             'file' => $exception->getFile(),
-            'line' => $exception->getLine()
+            'line' => $exception->getLine(),
         ];
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:49]: can we also get a column number?

File: src/Util/Stacktrace.php [hunk 2]
```diff
@@ -35,18 +35,32 @@ public static function buildStack(Throwable $exception): array
...
-         * Reverse stack to go through from first call in project's entry point
+         * Frames iterator
+         */
+        $i = 0;
+
+        /**
+         * Add real error's path to trace chain
          */
-        $stack = array_reverse($stack);
+        $newStack[$i] = [
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:59]: I see same [lines](https://github.com/codex-team/hawk.php/pull/30/files#diff-74543a71ba049e8ca882af5cb1bcfe0fdf14071f481d062e8cff49bdcfda1321R119-R123) below. Maybe we can remove duplication

File: src/Util/Stacktrace.php [hunk 3]
```diff
@@ -97,32 +111,152 @@ public static function buildStack(Throwable $exception): array
...
              */
-            $stack[$index]['trace'] = self::getAdjacentLines($callee['file'], $callee['line']);
+            $newStack[++$i] = [
+                'file'       => $frame['file'],
+                'line'       => $frame['line'],
+                'sourceCode' => self::getAdjacentLines($frame['file'], $frame['line']),
+            ];
+
+            /**
+             * Fill function and arguments data for the previous frame
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:126]: why previous? describe in comments

File: src/Util/Stacktrace.php [hunk 4]
```diff
@@ -97,32 +111,152 @@ public static function buildStack(Throwable $exception): array
...
+            $newStack[$i - 1]['arguments'] = self::getArgs($frame);
         }
 
+        return $newStack;
+    }
+
+    /**
+     * Compose function name with a class
+     *
+     * @param $frame
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:138]: no type, no description

File: src/Util/Stacktrace.php [hunk 5]
```diff
@@ -97,32 +111,152 @@ public static function buildStack(Throwable $exception): array
...
+     *
+     * @return string
+     */
+    private static function composeFunctionName($frame)
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:142]: specify return and param types

File: src/Util/Stacktrace.php [hunk 6]
```diff
@@ -97,32 +111,152 @@ public static function buildStack(Throwable $exception): array
...
+             */
+            $newStack[$i - 1]['function'] = self::composeFunctionName($frame);
+            $newStack[$i - 1]['arguments'] = self::getArgs($frame);
         }
 
+        return $newStack;
+    }
+
+    /**
+     * Compose function name with a class
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:136]: What does it mean "with a class"? Add an example

File: src/Util/Stacktrace.php [hunk 7]
```diff
@@ -97,32 +111,152 @@ public static function buildStack(Throwable $exception): array
...
-        ];
+        $functionName = '';
 
         /**
-         * Normalize array's indexes
+         * Try to fill name with a class name and type '::' or '->'
          */
-        $stack = array_values($stack);
+        try {
+            if (isset($frame['class'])) {
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:153]: maybe `!empty()` ?

File: src/Util/Stacktrace.php [hunk 8]
```diff
@@ -97,32 +111,152 @@ public static function buildStack(Throwable $exception): array
...
+                $functionName = $frame['class'] . $frame['type'];
+            }
+        } catch (\Exception $e) {
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:156]: describe this case

File: src/Util/Stacktrace.php [hunk 9]
```diff
@@ -97,32 +111,152 @@ public static function buildStack(Throwable $exception): array
...
+    }
+
+    /**
+     * Get function arguments for a frame
+     *
+     * @param $frame
+     *
+     * @return array
+     */
+    private static function getArgs($frame)
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:174]: add param and return types

File: src/Util/Stacktrace.php [hunk 10]
```diff
@@ -97,32 +111,152 @@ public static function buildStack(Throwable $exception): array
...
+    }
+
+    /**
+     * Trying to create a reflection method
+     *
+     * @param $frame - backtrace frame
+     *
+     * @return \ReflectionFunction|\ReflectionMethod|null
+     */
+    private static function getReflectionMethod($frame)
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:233]: specify types

File: src/Util/Stacktrace.php [hunk 11]
```diff
@@ -97,32 +111,152 @@ public static function buildStack(Throwable $exception): array
...
+         * ReflectionFunction/ReflectionMethod class reports information
+         * about a function/method.
+         */
+        $reflection = self::getReflectionMethod($frame);
+
+        /**
+         * If reflection function in missing then create a simple list of arguments
+         */
+        if (!$reflection) {
+            foreach ($frame['args'] as $index => $value) {
```
Comment (neSpecc) [inline src/Util/Stacktrace.php:191]: unsafe calling `$frame['args']`. Is it always specified?

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: src/Util/Stacktrace.php (hunk 1 above)
Comment (talyguryn) [reply src/Util/Stacktrace.php:49]: no. php exception does not contain this info

File: src/Util/Stacktrace.php (hunk 10 above)
Comment (talyguryn) [reply src/Util/Stacktrace.php:233]: as i know, we can not use multiple types for a function return for php <8

File: src/Util/Stacktrace.php (hunk 11 above)
Comment (talyguryn) [reply src/Util/Stacktrace.php:191]: yep. but i'll add a check above

File: src/Util/Stacktrace.php (hunk 3 above)
Comment (talyguryn) [reply src/Util/Stacktrace.php:124]: done

File: src/Util/Stacktrace.php (hunk 2 above)
Comment (talyguryn) [reply src/Util/Stacktrace.php:57]: done

Comment (neSpecc) [reply src/Util/Stacktrace.php:57]: it is not done

Comment (talyguryn) [reply src/Util/Stacktrace.php:57]: oh, i see. it doesn't matter. this part always will be, but that part is in a cycle. this is the simplest solution

File: src/Util/Stacktrace.php [hunk 12]
```diff
@@ -97,39 +109,174 @@ public static function buildStack(Throwable $exception): array
...
+    }
+
+    /**
+     * Compose function name with a class for frame
+     *
+     * @param array $frame - backtrace frame
+     *
+     * @return string
+     */
+    private static function composeFunctionName($frame): string
```
Comment (khaydarov) [inline src/Util/Stacktrace.php:150]:
```suggestion
    private static function composeFunctionName(array $frame): string
```

File: src/Util/Stacktrace.php [hunk 13]
```diff
@@ -97,39 +109,174 @@ public static function buildStack(Throwable $exception): array
...
+    }
+
+    /**
+     * Get function arguments for a frame
+     *
+     * @param array $frame - backtrace frame
+     *
+     * @return array
+     */
+    private static function getArgs($frame): array
```
Comment (khaydarov) [inline src/Util/Stacktrace.php:179]:
```suggestion
    private static function getArgs(array $frame): array
```

File: src/Util/Stacktrace.php [hunk 14]
```diff
@@ -97,39 +109,174 @@ public static function buildStack(Throwable $exception): array
...
+    }
+
+    /**
+     * Trying to create a reflection method
+     *
+     * @param array $frame - backtrace frame
+     *
+     * @return \ReflectionFunction|\ReflectionMethod|null
+     */
+    private static function getReflectionMethod($frame)
```
Comment (khaydarov) [inline src/Util/Stacktrace.php:246]:
```suggestion
    private static function getReflectionMethod(array $frame): ?ReflectionFunctionAbstract
```

File: src/Util/Stacktrace.php [hunk 15]
```diff
@@ -35,18 +35,30 @@ public static function buildStack(Throwable $exception): array
          */
         $stack = $exception->getTrace();
 
+        var_dump($stack);
```
Comment (khaydarov) [inline src/Util/Stacktrace.php:38]: dont forget to remove pls

**Review by khaydarov — APPROVED** [review]

**Review by neSpecc — APPROVED** [review]

**Pull request merged**